src/persist-client/src/operators/shard_source.rs

```diff
@@ -0,0 +1,467 @@
+            .open_leased_reader::<K, V, T, D>(shard_id, &format!("shard_source({})", name_owned))
+            .await
+            .expect("could not open persist shard");
+        let as_of = as_of.unwrap_or_else(|| read.since().clone());
+
+        // Eagerly yield the initial as_of. This makes sure that the output
+        // frontier of the `persist_source` closely tracks the `upper` frontier
+        // of the persist shard. It might be that the snapshot for `as_of` is
+        // not initially available yet, but this makes sure we already downgrade
+        // to it.
+        //
+        // Downstream consumers might rely on close frontier tracking for making
+        // progress. For example, the `persist_sink` needs to know the
+        // up-to-date upper of the output shard to make progress because it
+        // will only write out new data once it knows that earlier writes went
+        // through, including the initial downgrade of the shard upper to the
+        // `as_of`.
+        yield (Vec::new(), as_of.clone(), 0);
+
+        let subscription = read.subscribe(as_of.clone()).await;
+
+        let mut subscription = subscription.unwrap_or_else(|e| {
+            panic!(
+                "{}: {} cannot serve requested as_of {:?}: {:?}",
+                name_owned, shard_id, as_of, e
+            )
+        });
+
+        let mut done = false;
+        while !done {
+            while let Ok(leased_part) = consumed_part_rx.try_recv() {
+                subscription
+                    .return_leased_part(subscription.leased_part_from_exchangeable(leased_part));
+            }
+
+            let (parts, progress) = subscription.next().await;
+            // If `until.less_equal(progress)`, it means that all subsequent batches will
+            // contain only times greater or equal to `until`, which means they can be dropped
+            // in their entirety. The current batch must be emitted, but we can stop afterwards.
+            if PartialOrder::less_equal(&until, &progress) {
+                done = true;
+            }
+
+            let parts_size_bytes = parts.iter().map(|x| x.encoded_size_bytes()).sum();
+            yield (parts, progress, parts_size_bytes);
+        }
+
+        // Rather than simply end, we spawn a task that can continue to return leases.
+        // This keeps the `ReadHandle` alive until we have confirmed each lease as read.
+        mz_ore::task::spawn(|| "LeaseReturner", async move {
+            while let Some(leased_part) = consumed_part_rx.recv().await {
+                subscription
+                    .return_leased_part(subscription.leased_part_from_exchangeable(leased_part));
+            }
+        });
+    });
+
+    let mut pinned_stream = Box::pin(async_stream);
+
+    let mut builder =
+        AsyncOperatorBuilder::new(format!("shard_source_descs({})", name), scope.clone());
+    let mut flow_control_input = builder.new_input(flow_control_input, Pipeline);
```

Comment:
Are we disconnecting this in the end? I was convinced by the conversation yesterday that we want this to be disconnected for maximum throughput

Reply:
done!